### DISCUSSION TITLE
How to abstract view-build code into a method

### DISCUSSION THREAD

**Discussion by blackwhite-bear** [hecrj/iced#1322] [General]
Because of the complex view that I want to build, but make them in a few separate parts, so I want to abstract it into a method like the below picture, there is a confusing problem. 

'store_view' is located into two methods, one of them as a part of 'Counter'. However, the compiler tells the issue 'the trait bound `iced_native::element::Element<'_, Message, iced_graphics::renderer::Renderer<iced_glow::backend::Backend>>: From<&mut iced_native::element::Element<'_, Message, iced_graphics::renderer::Renderer<iced_glow::backend::Backend>>>` is not satisfied
the following implementations were found:'

Why does it happen?

![截图_滚动截图_20220428122852](https://user-images.githubusercontent.com/80373863/165676787-65e46da0-4711-4f6d-9177-6cdd4f949fff.png)

**Comment by genusistimelord** [comment]
this happens because the first store_view takes a mutable reference into the scrollable while the 2nd one also takes and holds a reference within the scrollable. try removing the first store_view().  I do something similar to this in my Iced Example of a item editor.

https://github.com/AscendingCreations/Iced_demo/blob/main/itemeditor/src/item/item_ui.rs

**Reply by blackwhite-bear** [reply]
I am not sure this issue is caused by the reference order. I move it to the first place calling but the issue happens there. It seems like there is something wrong with the return type.

Actually, I have looked into your example. It is great. However, the reason I want to abstract some UI-building logic is that it can be very long. So, I want to split them into a few parts by their interaction logic. Even following the pattern like yours, they are still together. Do you have new ideas?

![image](https://user-images.githubusercontent.com/80373863/165752257-530c3994-7fce-4898-a6f0-4d36c57b756e.png)

**Reply by genusistimelord** [reply]
what is the error message it is currently giving you after you moved it?

**Reply by genusistimelord** [reply]
This issue should be a lot easier to solve with the Pure widgets that will get released with 0.4. since they no longer require you to hold onto the state.

Anyways try passing the Element<_>  without the &mut to push.

**Reply by blackwhite-bear** [reply]
the trait bound `iced_native::element::Element<'_, _, iced_graphics::renderer::Renderer<iced_glow::backend::Backend>>: From<&mut iced_native::element::Element<'_, Message, iced_graphics::renderer::Renderer<iced_glow::backend::Backend>>>` is not satisfied
the following implementations were found:
  <iced_native::element::Element<'a, Message, Renderer> as From<Image>>
  <iced_native::element::Element<'a, Message, Renderer> as From<Svg>>
  <iced_native::element::Element<'a, Message, Renderer> as From<Viewer<'a>>>
  <iced_native::element::Element<'a, Message, Renderer> as From<iced::Space>>
and 14 others
required because of the requirements on the impl of `Into<iced_native::element::Element<'_, _, iced_graphics::renderer::Renderer<iced_glow::backend::Backend>>>` for `&mut iced_native::element::Element<'_, Message, iced_graphics::renderer::Renderer<iced_glow::backend::Backend>>`rustcE0277
main.rs(69, 18): required by a bound introduced by this call
row.rs(96, 12): required by a bound in `iced_native::widget::row::Row::<'a, Message, Renderer>::push`

**Reply by genusistimelord** [reply]
change it from  `&mut self.store_view()`  to `self.store_view()` As there is no Into for &mut element

**Reply by blackwhite-bear** [reply]
Do you mean like this?

![image](https://user-images.githubusercontent.com/80373863/165756415-7c4df57f-47b1-4305-89b5-0d7bbee9da8e.png)

**Reply by genusistimelord** [reply]
you can leave the Message within the Element. I only used _ as a short hand. So re-add that back and you should be fine.

**Reply by genusistimelord** [reply]
is it now fixed for you?

**Reply by blackwhite-bear** [reply]
I am not sure I have gotten your idea right. Here are a few situations I have made.

- self has to remain the mutable property, otherwise, we can not pass a mutable state for a widget
- I can not mark the return type as Element<_>, the compiler wants me to indicate the type

![issue](https://user-images.githubusercontent.com/80373863/165759854-d11d4a1e-7f00-4efd-b744-7965f2fad703.jpg)

**Reply by genusistimelord** [reply]
I never said to change the `Element` to be `Element<_>`. Change it back to being  `Element<Message>`
I only want you to not make the `&mut self.store_view()`  into  `self.store_view()`  this is fine as even if store_view requires a &mut self the Self will set itself as &mut within the store_view().  

^ do the above \/ Below is the Explanation as to why this is occurring.

when you are setting &mut in front of a Function call the &mut is used against the Return type so in this case since the Return is `Element<Message>`  you are placing  ` &mut Element<Message>` into the push function. So removing the &mut will solve that issue since `&mut Element<Message>` does not have Into implemented for it.

**Reply by blackwhite-bear** [reply]
- I marked the return type as Element, It does not work
- Do I removed the tag mut right?  

![2](https://user-images.githubusercontent.com/80373863/165765100-20887a78-9080-4ce2-9c45-a71305c44f93.jpg)

**Reply by genusistimelord** [reply]
* I marked the return type as Element, It does not work
    **_It has to be Element<Message>_**
* Do I removed the tag mut right?
   **_No you leave the Functions Argument as &mut self as the state for self.scroll must have mutability._**


Please send the Code as Code and not as an Image.

**Reply by blackwhite-bear** [reply]
Like this？

``` rust
impl Sandbox for Counter {
    type Message = Message;

    fn new() -> Self {
        Self {
            host_address: String::new(),
            value: 0,
            name_value: String::new(),
            increment_button: Default::default(),
            decrement_button: Default::default(),
            name_field: Default::default(),
            scroll: Default::default(),
        }
    }

    fn title(&self) -> String {
        String::from("Counter")
    }

    fn update(&mut self, message: Message) {
        match message {
            Message::IncrementPressed => {
                self.value += 1;
            }
            Message::DecrementPressed => {
                self.value -= 1;
            }
            Message::InputChanged(content) => {
                self.name_value = content;
            }
        }
    }

    fn view(&mut self) -> Element<Message> {
        self.build_view()
    }
}

impl Counter {
    fn build_view(&self) -> Element<Message> {
        Container::new(
            Row::new()
                .padding(10)
                .spacing(10)
                .push(self.store_view())
                .push(self.store_view()),
        )
        .width(Length::Fill)
        .height(Length::Fill)
        .style(light::Body)
        .into()
    }

    fn store_view(&mut self) -> Element {
        Container::new(
            Column::new().push(Text::new("list").size(30)).push(
                Scrollable::new(&mut self.scroll)
                    .width(Length::Fill)
                    .spacing(10)
                    .push(Text::new("ABCDF")),
            ),
        )
        .width(Length::Units(280))
        .height(Length::Fill)
        .style(light::Container)
        .into()
    }
}```

**Reply by genusistimelord** [reply]
Try the below I made a change to it.
```rust
impl Sandbox for Counter {
    type Message = Message;

    fn new() -> Self {
        Self {
            host_address: String::new(),
            value: 0,
            name_value: String::new(),
            increment_button: Default::default(),
            decrement_button: Default::default(),
            name_field: Default::default(),
            scroll: Default::default(),
        }
    }

    fn title(&self) -> String {
        String::from("Counter")
    }

    fn update(&mut self, message: Message) {
        match message {
            Message::IncrementPressed => {
                self.value += 1;
            }
            Message::DecrementPressed => {
                self.value -= 1;
            }
            Message::InputChanged(content) => {
                self.name_value = content;
            }
        }
    }

    fn view(&mut self) -> Element<Message> {
        self.build_view()
    }
}

impl Counter {
    fn build_view(&mut self) -> Element<Message> {
        Container::new(
            Row::new()
                .padding(10)
                .spacing(10)
                .push(self.store_view())
                .push(self.store_view()),
        )
        .width(Length::Fill)
        .height(Length::Fill)
        .style(light::Body)
        .into()
    }

    fn store_view(&mut self) -> Element<Message> {
        Container::new(
            Column::new().push(Text::new("list").size(30)).push(
                Scrollable::new(&mut self.scroll)
                    .width(Length::Fill)
                    .spacing(10)
                    .push(Text::new("ABCDF")),
            ),
        )
        .width(Length::Units(280))
        .height(Length::Fill)
        .style(light::Container)
        .into()
    }
}
```

**Reply by genusistimelord** [reply]
the only issue now is  you cant reference scroll twice to 2 different things. So in order to split this you will need to have 2 separate scroll variables.

**Reply by blackwhite-bear** [reply]
I have made this change, and then……
made another function `table_info` to create a collection of widgets and there is no reference to the widgets the first function uses. However, the error is still there. 


``` rust

use iced::{button, scrollable, text_input, window};
use iced::{Column, Length, Row, Settings};
use iced::{Container, Element, Sandbox, Scrollable, Text, TextInput};

// mod theme;
// use theme::light;

struct Counter {
    host_address: String,
    value: i32,
    name_value: String,
    increment_button: button::State,
    decrement_button: button::State,
    name_field: text_input::State,
    scroll: scrollable::State,
}

#[derive(Debug, Clone)]
enum Message {
    IncrementPressed,
    DecrementPressed,
    InputChanged(String),
}

impl Sandbox for Counter {
    type Message = Message;

    fn new() -> Self {
        Self {
            host_address: String::new(),
            value: 0,
            name_value: String::new(),
            increment_button: Default::default(),
            decrement_button: Default::default(),
            name_field: Default::default(),
            scroll: Default::default(),
        }
    }

    fn title(&self) -> String {
        String::from("Counter")
    }

    fn update(&mut self, message: Message) {
        match message {
            Message::IncrementPressed => {
                self.value += 1;
            }
            Message::DecrementPressed => {
                self.value -= 1;
            }
            Message::InputChanged(content) => {
                self.name_value = content;
            }
        }
    }

    fn view(&mut self) -> Element<Message> {
        self.build_view()
    }
}

impl Counter {
    fn build_view(&mut self) -> Element<Message> {
        Container::new(
            Row::new()
                .padding(10)
                .spacing(10)
                .push(self.store_view())
                .push(self.table_info()),
        )
        .width(Length::Fill)
        .height(Length::Fill)
        // .style(light::Body)
        .into()
    }

    fn store_view(&mut self) -> Element<Message> {
        Container::new(
            Column::new().push(Text::new("list").size(30)).push(
                Scrollable::new(&mut self.scroll)
                    .width(Length::Fill)
                    .spacing(10)
                    .push(Text::new("ABCDF")),
            ),
        )
        .width(Length::Units(280))
        .height(Length::Fill)
        // .style(light::Container)
        .into()
    }

    fn table_info(&mut self) -> Element<Message> {
        Column::new()
            .push(Text::new(&self.host_address))
            .push(
                Row::new()
                    .spacing(8)
                    .push(Text::new("address: "))
                    .push(
                        TextInput::new(&mut self.name_field, "", "", Message::InputChanged)
                            .width(Length::Fill)
                            .padding(5),
                    )
                    .height(Length::Units(28)),
            )
            .spacing(10)
            .width(Length::Fill)
            .height(Length::Fill)
            .into()
    }
}

fn main() -> iced::Result {
    Counter::run(Settings {
        window: window::Settings {
            size: (800, 600),
            resizable: true,
            min_size: (500, 500).into(),

            ..window::Settings::default()
        },

        default_font: Some(include_bytes!("NotoSerifCJK-Regular.ttc")),
        ..Settings::default()
    })
}
```
